### PR TITLE
[tt-train] Refactor: Move common utilities up one level and unify dataflow operations

### PR DESCRIPTION
**Pull request by mdragulaTT**
### Ticket
[[tt-train] Move tile read/write utilities to common dataflow utils](https://github.com/tenstorrent/tt-metal/issues/31125)

### Problem description
Tile read/write utilities were duplicated across multiple dataflow kernels (SwiGLU, SiLU, RMSNorm, LayerNorm, Cross Entropy, SDPA, Softmax, SGD Fused, Profiler No-Op). Common utilities were scattered in `ops/common/` and device operations used inconsistent `static_cast<int>` instead of proper `enchantum::to_string` for enum validation.

### What's changed
- **Moved common utilities**: `sources/ttml/metal/ops/common/` → `sources/ttml/metal/common/`
- **Added templated dataflow utilities**: `read_tiles_by_row()`, `read_tiles_by_col()`, `write_tiles_by_row()` in `dataflow_utils.hpp`
- **Modernized NOC API**: Updated from `noc_async_read_tile`/`noc_async_write_tile` to `noc_async_read_page`/`noc_async_write_page`
- **Replaced duplicated code**: Updated all dataflow kernels to use shared utilities, removing duplicated code
- **Fixed enum conversions**: Replaced `static_cast<int>` with `enchantum::to_string` in device operations
- **Standardized include paths**: Full paths for kernels, relative paths for program factories

**Impact**: Single source of truth for tile access patterns, improved maintainability, consistent NOC API usage across all operations.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/dataflow/reader_rmsnorm_bw_interleaved_start_id.cpp [hunk 1]
```diff
@@ -85,34 +69,33 @@ void kernel_main() {
...
+            // Barrier called by read_tiles_by_row with UseBarrier=true above
             cb_push_back(cb_input_idx, block_size);
             cb_push_back(cb_dL_out_idx, block_size);
-#ifndef EVERYTHING_FITS_IN_L1
-            cb_push_back(cb_gamma_idx, block_size);
 #else
+            // If everything fits in L1, we can read gamma only once.
             if (r == start_row) {
-                cb_push_back(cb_gamma_idx, block_size);
+                read_tiles_by_row(cb_gamma_idx, gamma_address_generator, c, block_size, tile_bytes, block_size);
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/kernels/dataflow/reader_rmsnorm_bw_interleaved_start_id.cpp:91]: Not true, push_back is happening inside read_tiles_by_row

File: tt-train/sources/ttml/metal/ops/silu_bw/device/kernels/dataflow/reader_silu_bw_interleaved_start_id.cpp [hunk 2]
```diff
@@ -62,12 +45,12 @@ void kernel_main() {
...
-            read_tiles(cb_input_idx, input_address_generator, row_tile_idx, block_size, current_block_size, tile_bytes);
-            read_tiles(
-                cb_dL_out_idx, dL_out_address_generator, row_tile_idx, block_size, current_block_size, tile_bytes);
-            noc_async_read_barrier();
+            read_tiles_by_row</* UseBarrier = */ false>(
+                cb_input_idx, input_address_generator, row_tile_idx, current_block_size, tile_bytes, block_size);
+            read_tiles_by_row(
+                cb_dL_out_idx, dL_out_address_generator, row_tile_idx, current_block_size, tile_bytes, block_size);
+            // Barrier called by read_tiles_by_row with UseBarrier=true above
             cb_push_back(cb_input_idx, block_size);
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/ops/silu_bw/device/kernels/dataflow/reader_silu_bw_interleaved_start_id.cpp:53]: Not true, push_back is happening inside read_tiles_by_row

File: tt-train/sources/ttml/CMakeLists.txt [hunk 3]
```diff
@@ -191,7 +191,9 @@ set(METAL_OPS_FILES
     ${CMAKE_CURRENT_SOURCE_DIR}/metal/operations.hpp
     ${CMAKE_CURRENT_SOURCE_DIR}/metal/ttnn_all_includes.hpp
     # Common
-    ${CMAKE_CURRENT_SOURCE_DIR}/metal/ops/common/program_utils.hpp
```
Comment (vmelnykovTT) [inline tt-train/sources/ttml/CMakeLists.txt:194]: Just to clarify — why was this moved out of the ops/ folder? My understanding is that these functions are only used by kernels.

File: tt-train/sources/ttml/metal/common/dataflow_utils.hpp [hunk 4]
```diff
@@ -156,6 +156,142 @@ inline float uint32_to_float(uint32_t bits) {
...
+    const uint32_t cb_idx,
+    const AddrGen& addr_gen,
+    const uint32_t start_idx,
+    const uint32_t num_tiles_to_read,
+    const uint32_t tile_bytes,
+    const uint32_t num_tiles_to_push) {
+    cb_reserve_back(cb_idx, num_tiles_to_push);
+    uint32_t l1_addr = get_write_ptr(cb_idx);
+    for (uint32_t t = 0; t < num_tiles_to_read; ++t) {
+        noc_async_read_page(start_idx + t, addr_gen, l1_addr);
```
Comment (vmelnykovTT) [inline tt-train/sources/ttml/metal/common/dataflow_utils.hpp:183]: Have you tested noc_async_read_page to confirm it works correctly?

Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/common/dataflow_utils.hpp:183]: I’m also wondering how the page size is determined — how does the function know it should read exactly one tile?

Additionally, what’s the maximum supported page size? For example, can we read 2 or 3 tiles in a single call?
Reading multiple tiles at once might be more efficient, so it would be cool to figure this out and maybe even add some optimization here depending on block_size.

As I can see from `tt_metal/hw/inc/dataflow_api.h` page_size is define by addrgen

```
uint32_t page_size;
if constexpr (has_page_size_v<AddrGen>) {
    page_size = addrgen.page_size;
} else {
    page_size = (1 << addrgen.log_base_2_of_page_size);
}
```
    
Maybe we can add this optimization by next PR, but we should investigate this

**Review by vmelnykovTT — APPROVED** [review]
Nice refactor and cleanup PR — great work!
Really appreciate the effort you put into this.
LGTM!

File: tt-train/sources/ttml/CMakeLists.txt (hunk 3 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/CMakeLists.txt:194]: Discussed offline, but long story short, those utils are also used in `SGDFused`

File: tt-train/sources/ttml/metal/common/dataflow_utils.hpp (hunk 4 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/metal/common/dataflow_utils.hpp:183]: Discussed offline. Seems like in `tile_layout`, the tiles can be spread across different banks so reading a few pages in one call is not possible, but maybe possible in row-major layour

**Comment by zbaczewskiTT** [comment]
Thank you! LGTM!

**Review by dyurshevichTT — APPROVED** [review]
Good job! Everything looks good to me!

**Review by dmakoviichuk-tt — APPROVED** [review]
Looks nice!
I like the direction where you are moving :)
